WikiWatershed/model-my-watershed
# Cache censuses

Cache AoI and modification censuses.
* If a AoI census or modification censuses have already been generated
and are still valid, skip the geoprocessing tasks and run TR55 directly.
* If only the AoI census is available, generate the modification
censuses and provide the cached AoI census to TR-55.
* Modify the structure of the Scenario.census to contain the
modification hash from when the censuses were generated, as well
as both the AoI census and the modification censuses in the structure
expected in the geoprocessing tasks.
* Update and add tests to test caching logic.

Connects to #870

**Testing instructions:**
- Start a new project with a fairly large AoI so that it's easy to see the difference in computation time.

![screen shot 2015-10-02 at 12 04 56 pm](https://cloud.githubusercontent.com/assets/1042475/10251311/d26c5ece-68fd-11e5-9636-3ebca69294d6.png)

- Switch to the "New Scenario"
- The initial model run should take a few seconds.
- Change just the precipitation slider.
- The model run should be faster than the previous run.
- Add a modification. The model run should be slower than the last run, and maybe a bit slower than the first run (more modifications means a longer model run time).
- Change just the precipitation slider again.
- The model run should be fast again.
- Repeat until you are confident there is some caching going on.
- Run the tests.

@jamesmcclain had a good idea that I didn't take on in this card because I spent a lot of time upfront reading through and rewriting things, and it's a discrete task. When a new scenario is added, we should pass along the AoI census from the previous or first scenario if the AoI census exists. Issue here: https://github.com/WikiWatershed/model-my-watershed/issues/897

Looking for input from @jamesmcclain re: new function and variable names, and function comments.
